packages/messaging/src/core/api.test.ts

```diff
@@ -0,0 +1,215 @@
+      compareHeaders(expectedHeaders, actualHeaders);
+    });
+
+    it('throws if there is a problem with the response', async () => {
+      fetchStub.rejects(new Error('Fetch failed'));
+      await expect(
+        requestGetToken(firebaseDependencies, tokenDetails.subscriptionOptions!)
+      ).to.be.rejectedWith('Fetch failed');
+
+      fetchStub.resolves(
+        new Response(JSON.stringify({ error: { message: 'error message' } }))
+      );
+      await expect(
+        requestGetToken(firebaseDependencies, tokenDetails.subscriptionOptions!)
+      ).to.be.rejectedWith('messaging/token-subscribe-failed');
+
+      fetchStub.resolves(
+        new Response(
+          JSON.stringify({
+            /* no token */
+          })
+        )
+      );
+      await expect(
+        requestGetToken(firebaseDependencies, tokenDetails.subscriptionOptions!)
+      ).to.be.rejectedWith('messaging/token-subscribe-no-token');
+    });
+  });
+
+  describe('updateToken', () => {
+    it('calls the updateRegistration server API with correct parameters', async () => {
+      fetchStub.resolves(
+        new Response(JSON.stringify({ token: 'fcm-token-from-server' }))
+      );
+
+      const response = await requestUpdateToken(
+        firebaseDependencies,
+        tokenDetails
+      );
+
+      const expectedHeaders = new Headers({
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+        'x-goog-api-key': 'apiKey',
+        'x-goog-firebase-installations-auth': `FIS authToken`
+      });
+      const expectedBody: ApiRequestBody = {
+        web: {
+          endpoint: 'https://example.org',
+          auth: 'YXV0aC12YWx1ZQ',
+          p256dh: 'cDI1Ni12YWx1ZQ',
+          applicationPubKey: 'dmFwaWQta2V5LXZhbHVl'
+        }
+      };
+      const expectedRequest: RequestInit = {
+        method: 'PATCH',
+        headers: expectedHeaders,
+        body: JSON.stringify(expectedBody)
+      };
+      const expectedEndpoint = `${ENDPOINT}/projects/projectId/registrations/token-value`;
+
+      expect(response).to.equal('fcm-token-from-server');
+      expect(fetchStub).to.be.calledOnceWith(expectedEndpoint, expectedRequest);
+      const actualHeaders = fetchStub.lastCall.lastArg.headers;
+      compareHeaders(expectedHeaders, actualHeaders);
+    });
+
+    it('throws if there is a problem with the response', async () => {
+      fetchStub.rejects(new Error('Fetch failed'));
+      await expect(
+        requestUpdateToken(firebaseDependencies, tokenDetails)
+      ).to.be.rejectedWith('Fetch failed');
+
+      fetchStub.resolves(
+        new Response(JSON.stringify({ error: { message: 'error message' } }))
+      );
+      await expect(
+        requestUpdateToken(firebaseDependencies, tokenDetails)
+      ).to.be.rejectedWith('messaging/token-update-failed');
+
+      fetchStub.resolves(
+        new Response(
+          JSON.stringify({
+            /* no token */
```

Comment:
You could replace these four lines with the four characters "{}" but I guess consistency is important too :stuck_out_tongue:

Reply:
Yeah that's true, but I think this way better expresses what I am actually doing here. Plus consistency, of course. :sweat_smile: